### PR TITLE
Suggest types for constants that are untyped

### PR DESCRIPTION
**Pull request by DarkDimius**
### Motivation
As many people pointed out, sorbet _does_ know types for untyped constants at some point.
This builds build a way to burn those types into source code.
This is build on prior exploration that @nelhage did.
### Test plan

See included automated tests.

cc @sushain97

### CODE REVIEW

File: core/types/calls.cc [hunk 1]
```diff
@@ -1567,6 +1567,23 @@ class Magic_callWithSplatAndBlock : public IntrinsicMethod {
...
 
+class Magic_suggestUntypedConstantType : public IntrinsicMethod {
+public:
+    void apply(Context ctx, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        ENFORCE(args.args.size() == 1);
+        auto ty = args.args.front()->type;
+        auto loc = args.locs.args[0];
+        if (!ty->isUntyped()) {
+            if (auto e = ctx.state.beginError(loc, core::errors::Infer::UntypedConstantSuggestion)) {
+                e.setHeader("Suggested type: `{}`", ty->show(ctx));
```
Comment (jez) [inline core/types/calls.cc]: Might be nice to connect this to the constant missing type error that will have already been reported above.

File: core/types/calls.cc [hunk 2]
```diff
@@ -1567,6 +1567,23 @@ class Magic_callWithSplatAndBlock : public IntrinsicMethod {
...
+                e.addAutocorrect(
+                    core::AutocorrectSuggestion{loc, fmt::format("T.let({}, {})", loc.source(ctx), ty->show(ctx))});
```
Comment (jez) [inline core/types/calls.cc]: nit: `e.replaceWith`

non-blocking i can do this later if you want

File: test/cli/suggest-constant-type/suggest-constant-type.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+cwd="$(pwd)"
+cd "$tmp" || exit 1
+
+"$cwd/main/sorbet" --silence-dev-message -a suggest-constant-type.rb 2>&1
+
+echo
+echo --------------------------------------------------------------------------
+echo
+
+# Also cat the file, to make sure that 'extend' is only added once per class.
```
Comment (jez) [inline test/cli/suggest-constant-type/suggest-constant-type.sh]: This comment doesn't apply

File: core/types/calls.cc (hunk 1 above)
Comment (DarkDimius) [reply core/types/calls.cc]: Ack, I don't see how to do this easily.
I might remove the error on the other end completely though...

**Pull request merged**